library/alloc/src/collections/btree/node.rs

```diff
@@ -1796,5 +1694,42 @@ unsafe fn slice_remove<T>(slice: &mut [MaybeUninit<T>], idx: usize) -> T {
     }
 }
 
+/// Shifts the slice such that the element at `mid` becomes the first.
+/// Is similar to `rotate_left`, but assumes the leading elements are
+/// uninitialized and leaves the trailing elements uninitialized.
+///
+/// # Safety
+/// The slice has more than `mid` elements.
+unsafe fn slice_align_left<T>(slice: &mut [MaybeUninit<T>], mid: usize) {
+    unsafe {
+        let slice_ptr = slice.as_mut_ptr();
+        ptr::copy(slice_ptr.add(mid), slice_ptr, slice.len() - mid);
+    }
+}
+
+/// Shifts the slice such that the first element appears at `mid`,
+/// Is similar to `rotate_right` but assumes the trailing elements are
+/// uninitialized and leaves the leading elements uninitialized.
+///
+/// # Safety
+/// The slice has more than `mid` elements.
+unsafe fn slice_align_right<T>(slice: &mut [MaybeUninit<T>], mid: usize) {
+    unsafe {
+        let slice_ptr = slice.as_mut_ptr();
+        ptr::copy(slice_ptr, slice_ptr.add(mid), slice.len() - mid);
+    }
+}
+
+/// Moves all elements in the slice elsewhere, leaving them all uninitialized.
+///
+/// # Safety
+/// The `dest` slice is at least as long as the `source` slice.
+unsafe fn slice_move<T>(source: &[MaybeUninit<T>], dest: &mut [MaybeUninit<T>]) {
```

Comment:
Can we note in the comment here that this is equivalent to copy_from_slice aside from avoiding bounds checking?

Reply:
and aside from not requiring T to be `Copy` and not requiring the destination to be the same length.